### PR TITLE
Timeout policy interruptable by default

### PR DESCRIPTION
**Pull request by kkowalski**
Interruptable true as a more sensible default.

I think this might be more the expected behavior, I certainly was surprised that I need to set additional flags to get a useful timeout. When I define a timeout of 1s and an execution(rpc) takes 10s, getting a response after 10s that the computation timed out hardly introduces resiliency to the code. I think it might be more sensible to make the timeout policy interruptable and cancellable by default with ability to disable it for some reason. I would be interested about the reasoning behind this default.

If this is reasonable I can finish this PR, we would need some changes to the API.

Issue supporting this:
#209

### CODE REVIEW

**Comment by whiskeysierra** [comment]
> When I define a timeout of 1s and an execution(rpc) takes 10s, getting a response after 10s that the computation timed out hardly introduces resiliency to the code.

That only happens for synchronous executions, correct?

**Comment by kkowalski** [comment]
In async withCancel is also needed to stop execution sooner.
```java
@Test
    public void failsafeTimeoutAsync() throws ExecutionException, InterruptedException {
        Timeout<Object> timeout = Timeout.of(Duration.ofSeconds(1));
        Failsafe.with(timeout).getAsync(() -> {
            Thread.sleep(10000);
            System.out.println("Doing stuff even though I should be dead by now");
            return "Done";
        }).get();
    }
```

**Comment by whiskeysierra** [comment]
> In async withCancel is also needed to stop execution early.

In async mode you need to differentiate between the future being completed and the original task finishing. The interrupt would only make the task finish earlier, but the future should be finished early in both cases (from my understanding).

**Comment by Tembrel** [comment]
Currently, the user has three options:
1. Attempt to cancel timed-out tasks using interruption: Call `withCancel(true)`.
1. Attempt to cancel timed-out tasks without interrupting: Call `withCancel(false)`.
1. __Don't__ attempt to cancel timed-out tasks: No call to `withCancel`.

This proposal __removes__ that last option: If you don't call `withCancel`, you would get the first behavior, i.e., as if you called `withCancel(true)`.

Furthermore, you should only use interruption on tasks that are designed for it (and in general tasks are __not__ designed for interruption), so this is not a good choice for a default.

The nice thing about the current state of affairs is that it reflects the API for cancelling a task manually:
```java
Future<?> task = ...;
...
task.cancel(false); // cancel but don't try to interrupt
```
Which makes this behavior easy to understand:
```
Timeout<?> timeout = Timeout.of(...).withCancel(false);
FailsafeExecutor<?> exec = Failsafe.with(timeout); // calls `cancel(false)` on timeout
```
TBH, the boolean argument to `Future.cancel` is an ugly compromise, but we've been stuck with it now for 16 years, so it's best to work with it. (Full disclosure: I served in the [expert group that designed this ugly compromise](https://jcp.org/en/jsr/detail?id=166).)

**Comment by kkowalski** [comment]
> This proposal removes that last option: If you don't call withCancel, you would get the first behavior, i.e., as if you called withCancel(true).

@Tembrel Those were the mentioned changes, to provide the 3rd option we would need something like withoutCancel() or withCancel(cancellable, interruptible).

>Furthermore, you should only use interruption on tasks that are designed for it (and in general tasks are not designed for interruption), so this is not a good choice for a default.

Why is interrupting in this context disadvantageous, is it performance ?

It's been a while since I was comparing resilience libraries but I think this was the default in Hystrix and resilience4j. 

I figured and it does make sense that the API wants to reflect Future API.

**Comment by whiskeysierra** [comment]
The third option is not calling the method at all.

**Comment by Tembrel** [comment]
I wrote:
> Furthermore, you should only use interruption on tasks that are designed for it (and in general tasks are **not** designed for interruption), so this is not a good choice for a default.

@kkowalski wrote:
> Why is interrupting in this context disadvantageous, is it performance ?

I'll make a stronger claim that __running__ tasks aren't in general cancellable, whether or not interruption is used. Unless your task never calls into foreign code (code that you don't control), you're at the mercy of the cancellation and interruption policies of the foreign code.

Tasks can be designed to be cancellable __without__ using interruption, by implementing `cancel(boolean)` in a task-specific way, typically by polling some task-specific field. (In that case, the implementation is free to ignore the boolean argument.) But this is not normally the case for tasks that are created by the Failsafe library itself.

Tasks __can__ be designed to be cancellable via interruption, but most aren't. In many tasks that are __not__ so designed an interrupt is likely to be harmless, but occasionally it can lead to failure to clean up properly, which can be very hard to diagnose. For example, a task might be calling into code that uses interruption for an entirely different purpose &mdash; not recommended, but not forbidden, either &mdash; in which case a stray interrupt might be misinterpreted. Also, a catch block of `Throwable` or `Exception` can end up masking an `InterruptedException` by treating it as a general exception, and subsequent exception handling might end up recording a task failure rather than the neutral fact of a cancellation.

Task execution machinery cannot tell whether or not it is safe to try to cancel a task by interrupting it, so we need to give it that information when requesting cancellation. The `cancel(interruptIfRunning)` method is undeniably ugly, but it does force the user to confront the question: Do I believe that it is safe to try to cancel this task through interruption?

All this is to say that even if we had a `withoutCancel` method to configure a `Timeout` instance, we still wouldn't want the default behavior (not calling any method) to be equivalent to `withCancel(true)`.

This pull request is to avoid having to add a single method call when building a timeout policy instance. I think that making the explicit extra call is important, regardless of the boolean argument, because it tells the reader that we expect that if the task hasn't started running yet, it's worthwhile ensuring that it doesn't run. If the boolean argument is true, it further tells the reader that we expect that it's safe to interrupt it and that that might actually have the effect of cancelling it cleanly earlier than it otherwise would.

A very long way of saying that I'm not in favor of accepting this pull request. If you find that you are commonly expressing timeout policies that call for cancellation, and you don't like the extra verbiage, why not create a static factory method to hide the extra few characters?


@whiskeysierra wrote:
> The third option is not calling the method at all.

Currently not calling `withCancel` has behavior that is distinct from calling it, but under this pull request not calling it would be equivalent to calling `withCancel(true)`. @kkowalski suggests adding either `withoutCancel` or `withCancel(cancellable, interruptible)`. I don't like either of those much; the second is not good because it is inconsistent with the one-arg version of the method.

(Aside: There's a minor spelling error in the Failsafe code, which uses "interrupt**a**ble". The latter is an acceptable alternative in prose, but the Java platform uses the former consistently.)

**Comment by jhalterman** [comment]
My thinking echoes @tembrel's [comments](https://github.com/jhalterman/failsafe/pull/226#issuecomment-580297661). Interrupting by default (`withCancel(true)`) could be somewhat dangerous since not all code is designed to be interrupted. This leaves the question of whether we should cancel by default (`withCancel(false)`).

Synchronous executions can't be cancelled, only interrupted. Async executions can be cancelled, but cancelling an async execution that has already started typically has no effect. 

It could be possible that when using a `Timeout` with a `RetryPolicy` (ex: `Failsafe.with(timeout, retryPolicy)`), the `Timeout` might be exceeded during a retry delay, in which case it might make sense to just cancel the execution by default rather than letting the retry proceed. I can't think of a good reason why we'd still want a retry to proceed after a Timeout has already happened.

Thoughts?

**Comment by kkowalski** [comment]
> why not create a static factory method to hide the extra few characters?

Sure, there are options to hide this.

>(Aside: There's a minor spelling error in the Failsafe code, which uses "interruptable". The latter is an acceptable alternative in prose, but the Java platform uses the former consistently.)

Thank you :) I was more presenting the shape of the solution to answer the comment rather than providing the exact interface.

After thinking about this deeper for a library which purely offers constructs for handling failures in arbitrary execution contexts not specifically for access to remote systems it make sense to behave this way by default. I might've assumed too much that the majority of people will be using this library as a wrapper for "small controlled" contexts making network calls. @Tembrel thanks for highlighting the potential problems.

The least I think @Tembrel's comments on this PR provide a valuable rationale behind this.

**Comment by jhalterman** [comment]
@kkowalski No worries, thanks for sparking this discussion since it does highlight whether cancellation (without interruption) should default to true. Do you use Failsafe/Diehard at CircleCI? If so, mind if I add CircleCI to the [users list](https://jodah.net/failsafe/whos-using/#companies--organizations)?

**Comment by kkowalski** [comment]
> Separately, do you use Failsafe/Diehard at CircleCI? If so, mind if I add CircleCI to the users list?

We use diehard for some of our integrations. I think it's fine, go ahead and add it.

**Comment by Tembrel** [comment]
@jhalterman wrote:
> That said, when using a `Timeout` with a `RetryPolicy`, there could be a pending execution, in which case it might make sense to always just cancel the execution by default rather than letting the retry proceed. I can't think of a good reason why we'd still want a retry to proceed after a Timeout has already happened.

Yeah, I can't come up with a realistic counterexample where there's a timeout and you know somehow (or rather, Failsafe knows) that one or more tasks are enqueued for execution and yet for some reason you __don't__ want to prevent them from running. The closest I can come is to imagine that in some cases preventing a pending task from running could be more expensive from letting it run, but that involves deliberately ignoring what I know to be true of the current Failsafe implementation and getting on a high horse about not restricting future implementations.

In other words, there's no reason not to remove pending tasks on timeout. The default behavior, therefore, should be to call `cancel(false)` on `Future<?>` instances on timeout. For completeness, I now think it __would__  be a good idea to offer a `withoutCancel` method, as @kkowalski suggested, to suppress this default behavior, though probably it wouldn't see much use. And I think it makes sense to continue to offer `withCancel(boolean)` to allow users to state their expectation that it is safe to use interruption to cancel a running task, even though `withCancel(false)` would then be redundantly confirming the default behavior.

**Comment by Tembrel** [comment]
Coincidentally, this just came in from Doug Lea:

http://cs.oswego.edu/pipermail/concurrency-interest/2020-February/017061.html

Seems relevant here and in some outstanding issues.

**Comment by jhalterman** [comment]
Opened #234 to discuss the issue related to cancellation. Since interruption will stay non-default for now, I'm closing this.
